# Data Wrangling on Umuzi Personality Scores.

## Overview of Notebook. 

*This notebook is about working with Personality scores data that was acquired from umuzi recruits.
*We are going to be merging data frames, filtering data frames and manipulation rows and columns in order to make the data useful.

### Import libraries that we shall use.

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

### Read in Personality scores csv file.

* Below we use pandas pd.read_csv method to read in the personality scores csv data into the notebook and assign the name per_scores to this data. 

In [2]:
per_scores = pd.read_csv("personality_scores.csv", delimiter=';')

* Below we use the .head() method to display the first (5) rows of the dataframe in order to have an idea of how our dataframe looks.

In [3]:
per_scores.head()

ID Section 5 of 6 [I am always prepared.]  \
0   0                                 (3, 5)   
1   1                                 (3, 5)   
2   2                                 (3, 5)   
3   3                                 (3, 5)   
4   4                                 (3, 3)   

  Section 5 of 6 [I am easily disturbed.]  \
0                                  (4, 5)   
1                                  (4, 5)   
2                                  (4, 3)   
3                                  (4, 5)   
4                                  (4, 5)   

  Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                             (3, 5)       
1                                             (3, 5)       
2                                             (3, 3)       
3                                             (3, 3)       
4                                             (3, 3)       

  Section 5 of 6 [I am full of ideas.]  \
0                               (5, 5)   
1                               (5, 5)   
2                               (5, 5)   
3                               (5, 5)   
4                               (5, 3)   

  Section 5 of 6 [I am interested in people.]  \
0                                      (2, 3)   
1                                      (2, 5)   
2                                      (2, 5)   
3                                      (2, 5)   
4                                      (2, 3)   

  Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                             (5, 3)        
1                                             (5, 3)        
2                                             (5, 5)        
3                                             (5, 3)        
4                                             (5, 3)        

  Section 5 of 6 [I am not interested in other people's problems.]  \
0                                             (2, 3)                 
1                                             (2, 5)                 
2                                             (2, 5)                 
3                                             (2, 3)                 
4                                             (2, 3)                 

  Section 5 of 6 [I am not really interested in others.]  \
0                                             (2, 5)       
1                                             (2, 5)       
2                                             (2, 5)       
3                                             (2, 3)       
4                                             (2, 3)       

  Section 5 of 6 [I am quick to understand things.]      ...        \
0                                            (5, 5)      ...         
1                                            (5, 5)      ...         
2                                            (5, 5)      ...         
3                                            (5, 3)      ...         
4                                            (5, 5)      ...         

  Unnamed: 60 Unnamed: 61 Unnamed: 62 Unnamed: 63 Unnamed: 64 Unnamed: 65  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   

  Unnamed: 66 Unnamed: 67 Unnamed: 68 IPIP_HIGH_RISK  
0         NaN         NaN         NaN            NaN  
1         NaN         NaN         NaN            NaN  
2         NaN         NaN         NaN            NaN  
3         NaN         NaN         NaN            NaN  
4         NaN         NaN         NaN            NaN  

[5 rows x 70 columns]

### Examine the data frame for duplicates (based on ID), and drop any duplicates that exist. 

* Below we use various methods to examine the per_scores data to detemine whether there are any duplicate values based on ID.

* Below we create an ids variable to determine whether there are duplicate values. No duplicate values are in the dataframe.

In [4]:
ids = per_scores["ID"]
per_scores[ids.isin(ids[ids.duplicated()])]

Empty DataFrame
Columns: [ID, Section 5 of 6 [I am always prepared.], Section 5 of 6 [I am easily disturbed.], Section 5 of 6 [I am exacting (demanding) in my work.], Section 5 of 6 [I am full of ideas.], Section 5 of 6 [I am interested in people.], Section 5 of 6 [I am not interested in abstract ideas.], Section 5 of 6 [I am not interested in other people's problems.], Section 5 of 6 [I am not really interested in others.], Section 5 of 6 [I am quick to understand things.], Section 5 of 6 [I am quiet around strangers.], Section 5 of 6 [I am relaxed most of the time.], Section 5 of 6 [I am the life of the party.], Section 5 of 6 [I change my mood a lot.], Section 5 of 6 [I do not have a good imagination.], Section 5 of 6 [I don't like to draw attention to myself.], Section 5 of 6 [I don't mind being the center of attention.], Section 5 of 6 [I don't talk a lot.], Section 5 of 6 [I feel comfortable around people.], Section 5 of 6 [I feel little concern for others.], Section 5 of 6 [I feel others' emotions.], Section 5 of 6 [I follow a schedule.], Section 5 of 6 [I get chores done right away.], Section 5 of 6 [I get irritated easily.], Section 5 of 6 [I get stressed out easily.], Section 5 of 6 [I get upset easily.], Section 5 of 6 [I have a rich vocabulary.], Section 5 of 6 [I have a soft (kind) heart.], Section 5 of 6 [I have a vivid imagination.], Section 5 of 6 [I have difficulty understanding abstract ideas.], Section 5 of 6 [I have excellent ideas.], Section 5 of 6 [I have frequent mood swings.], Section 5 of 6 [I have little to say.], Section 5 of 6 [I insult people.], Section 5 of 6 [I keep in the background.], Section 5 of 6 [I leave my belongings lying around.], Section 5 of 6 [I like order.], Section 5 of 6 [I make a mess of things.], Section 5 of 6 [I make people feel at ease.], Section 5 of 6 [I neglect my duties.], Section 5 of 6 [I often feel blue (down).], Section 5 of 6 [I often forget to put things back in their proper place], Section 5 of 6 [I pay attention to details.], Section 5 of 6 [I seldom feel blue (down).], Section 5 of 6 [I spend time reflecting on things.], Section 5 of 6 [I start conversations.], Section 5 of 6 [I sympathize with others' feelings.], Section 5 of 6 [I take time out for others.], Section 5 of 6 [I talk to a lot of different people at parties.], Section 5 of 6 [I use difficult words.], Section 5 of 6 [I worry about things.], Unnamed: 51, Unnamed: 52, Unnamed: 53, Unnamed: 54, Unnamed: 55, Unnamed: 56, Unnamed: 57, Unnamed: 58, Unnamed: 59, Unnamed: 60, Unnamed: 61, Unnamed: 62, Unnamed: 63, Unnamed: 64, Unnamed: 65, Unnamed: 66, Unnamed: 67, Unnamed: 68, IPIP_HIGH_RISK]
Index: []

[0 rows x 70 columns]

* we get an output of 0 rows * 70 columns meaning there are zero duplicates.

In [5]:
per_scores[per_scores.duplicated(["ID"], keep=False)]

Empty DataFrame
Columns: [ID, Section 5 of 6 [I am always prepared.], Section 5 of 6 [I am easily disturbed.], Section 5 of 6 [I am exacting (demanding) in my work.], Section 5 of 6 [I am full of ideas.], Section 5 of 6 [I am interested in people.], Section 5 of 6 [I am not interested in abstract ideas.], Section 5 of 6 [I am not interested in other people's problems.], Section 5 of 6 [I am not really interested in others.], Section 5 of 6 [I am quick to understand things.], Section 5 of 6 [I am quiet around strangers.], Section 5 of 6 [I am relaxed most of the time.], Section 5 of 6 [I am the life of the party.], Section 5 of 6 [I change my mood a lot.], Section 5 of 6 [I do not have a good imagination.], Section 5 of 6 [I don't like to draw attention to myself.], Section 5 of 6 [I don't mind being the center of attention.], Section 5 of 6 [I don't talk a lot.], Section 5 of 6 [I feel comfortable around people.], Section 5 of 6 [I feel little concern for others.], Section 5 of 6 [I feel others' emotions.], Section 5 of 6 [I follow a schedule.], Section 5 of 6 [I get chores done right away.], Section 5 of 6 [I get irritated easily.], Section 5 of 6 [I get stressed out easily.], Section 5 of 6 [I get upset easily.], Section 5 of 6 [I have a rich vocabulary.], Section 5 of 6 [I have a soft (kind) heart.], Section 5 of 6 [I have a vivid imagination.], Section 5 of 6 [I have difficulty understanding abstract ideas.], Section 5 of 6 [I have excellent ideas.], Section 5 of 6 [I have frequent mood swings.], Section 5 of 6 [I have little to say.], Section 5 of 6 [I insult people.], Section 5 of 6 [I keep in the background.], Section 5 of 6 [I leave my belongings lying around.], Section 5 of 6 [I like order.], Section 5 of 6 [I make a mess of things.], Section 5 of 6 [I make people feel at ease.], Section 5 of 6 [I neglect my duties.], Section 5 of 6 [I often feel blue (down).], Section 5 of 6 [I often forget to put things back in their proper place], Section 5 of 6 [I pay attention to details.], Section 5 of 6 [I seldom feel blue (down).], Section 5 of 6 [I spend time reflecting on things.], Section 5 of 6 [I start conversations.], Section 5 of 6 [I sympathize with others' feelings.], Section 5 of 6 [I take time out for others.], Section 5 of 6 [I talk to a lot of different people at parties.], Section 5 of 6 [I use difficult words.], Section 5 of 6 [I worry about things.], Unnamed: 51, Unnamed: 52, Unnamed: 53, Unnamed: 54, Unnamed: 55, Unnamed: 56, Unnamed: 57, Unnamed: 58, Unnamed: 59, Unnamed: 60, Unnamed: 61, Unnamed: 62, Unnamed: 63, Unnamed: 64, Unnamed: 65, Unnamed: 66, Unnamed: 67, Unnamed: 68, IPIP_HIGH_RISK]
Index: []

[0 rows x 70 columns]

* Above the p_scores dataframe shows there are no duplicate values.

In [6]:
len(per_scores['ID'])-len(per_scores['ID'])

0

* Above we use the " len(per_scores['ID'])-len(per_scores['ID']) " to determine whether there are duplicate values, we get a 0, meaning that there are no duplicate entries in the ID column.

In [7]:
len(per_scores['ID'])-len(per_scores['ID'].drop_duplicates(keep='first'))

0

* Above we use the " len(per_scores['ID'])-len(per_scores['ID'].drop_duplicates(keep='first')) " to determine whether there are duplicate values to be dropped, we get a 0, meaning that there are no duplicate entries in the ID column that can be dropped.

In [8]:
per_scores['ID'].is_unique

True

In [9]:
per_scores.head(1)

ID Section 5 of 6 [I am always prepared.]  \
0   0                                 (3, 5)   

  Section 5 of 6 [I am easily disturbed.]  \
0                                  (4, 5)   

  Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                             (3, 5)       

  Section 5 of 6 [I am full of ideas.]  \
0                               (5, 5)   

  Section 5 of 6 [I am interested in people.]  \
0                                      (2, 3)   

  Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                             (5, 3)        

  Section 5 of 6 [I am not interested in other people's problems.]  \
0                                             (2, 3)                 

  Section 5 of 6 [I am not really interested in others.]  \
0                                             (2, 5)       

  Section 5 of 6 [I am quick to understand things.]      ...        \
0                                            (5, 5)      ...         

  Unnamed: 60 Unnamed: 61 Unnamed: 62 Unnamed: 63 Unnamed: 64 Unnamed: 65  \
0         NaN         NaN         NaN         NaN         NaN         NaN   

  Unnamed: 66 Unnamed: 67 Unnamed: 68 IPIP_HIGH_RISK  
0         NaN         NaN         NaN            NaN  

[1 rows x 70 columns]

* Above we use "per_scores['ID'].is_unique" to further show that there are no duplicated entries in the personality scores dataframe when using the ID column.

### Assert that the new data frame is the length of the unique entries of the original data frame. 

* First we create the new dataframe which we shall call per_scores1 and set the ID colum as the index. 

In [10]:
per_scores1 = per_scores.set_index('ID')

* Below we use the .tail() method to see the last five row of the per_score1 dataframe.

In [11]:
per_scores1.tail()

Section 5 of 6 [I am always prepared.]  \
ID                                            
1550                                 (3, 5)   
1551                                 (3, 3)   
1552                                 (3, 5)   
1553                                 (3, 5)   
1554                                 (3, 3)   

     Section 5 of 6 [I am easily disturbed.]  \
ID                                             
1550                                  (4, 5)   
1551                                  (4, 5)   
1552                                  (4, 3)   
1553                                  (4, 5)   
1554                                  (4, 5)   

     Section 5 of 6 [I am exacting (demanding) in my work.]  \
ID                                                            
1550                                             (3, 1)       
1551                                             (3, 5)       
1552                                             (3, 5)       
1553                                             (3, 5)       
1554                                             (3, 1)       

     Section 5 of 6 [I am full of ideas.]  \
ID                                          
1550                               (5, 5)   
1551                               (5, 3)   
1552                               (5, 5)   
1553                               (5, 5)   
1554                               (5, 5)   

     Section 5 of 6 [I am interested in people.]  \
ID                                                 
1550                                      (2, 5)   
1551                                      (2, 5)   
1552                                      (2, 5)   
1553                                      (2, 5)   
1554                                      (2, 5)   

     Section 5 of 6 [I am not interested in abstract ideas.]  \
ID                                                             
1550                                             (5, 5)        
1551                                             (5, 3)        
1552                                             (5, 5)        
1553                                             (5, 5)        
1554                                             (5, 3)        

     Section 5 of 6 [I am not interested in other people's problems.]  \
ID                                                                      
1550                                             (2, 5)                 
1551                                             (2, 3)                 
1552                                             (2, 3)                 
1553                                             (2, 5)                 
1554                                             (2, 3)                 

     Section 5 of 6 [I am not really interested in others.]  \
ID                                                            
1550                                             (2, 5)       
1551                                             (2, 5)       
1552                                             (2, 3)       
1553                                             (2, 5)       
1554                                             (2, 5)       

     Section 5 of 6 [I am quick to understand things.]  \
ID                                                       
1550                                            (5, 5)   
1551                                            (5, 5)   
1552                                            (5, 5)   
1553                                            (5, 5)   
1554                                            (5, 3)   

     Section 5 of 6 [I am quiet around strangers.]      ...       Unnamed: 60  \
ID                                                      ...                     
1550                                        (1, 3)      ...               NaN   
1551                                        (1, 5)      ...               NaN   
1552                                        (1, 3)      ...               NaN   


* Below we check out the shapes of per_scores and per_scores1 dataframe to compare them.

In [12]:
print(per_scores.shape)
print(per_scores1.shape)

(1555, 70)
(1555, 69)


* from the abouve output, we can tell that both dataframes have equal number of rows (1555) and per_score has has 70 columns while per_scores1 has 69 because we made the ID column the new index column.

In [13]:
#Use len to get the number of rows of per_scores
print(len(per_scores))
#Use len to get the number of rows of per_scores1
print(len(per_scores1))

1555
1555


### 2. Write functions that will (1) convert the answer on each of the personality test questions to a number, as set out in scoring.md, and (2) calculate total scores for each of the subscales (conscientiousness, emotional stability, openness to new experience, agreeableness, extraversion).  

* Below we create per_scores2 df and drop all NaN values.

In [14]:
per_scores2 = per_scores1.dropna(axis = 1)

* Below we use the .head() method to display the first (5) rows of the dataframe in order to have an idea of how our per_scores dataframe looks.

In [15]:
per_scores2.head()

Section 5 of 6 [I am always prepared.]  \
ID                                          
0                                  (3, 5)   
1                                  (3, 5)   
2                                  (3, 5)   
3                                  (3, 5)   
4                                  (3, 3)   

   Section 5 of 6 [I am easily disturbed.]  \
ID                                           
0                                   (4, 5)   
1                                   (4, 5)   
2                                   (4, 3)   
3                                   (4, 5)   
4                                   (4, 5)   

   Section 5 of 6 [I am exacting (demanding) in my work.]  \
ID                                                          
0                                              (3, 5)       
1                                              (3, 5)       
2                                              (3, 3)       
3                                              (3, 3)       
4                                              (3, 3)       

   Section 5 of 6 [I am full of ideas.]  \
ID                                        
0                                (5, 5)   
1                                (5, 5)   
2                                (5, 5)   
3                                (5, 5)   
4                                (5, 3)   

   Section 5 of 6 [I am interested in people.]  \
ID                                               
0                                       (2, 3)   
1                                       (2, 5)   
2                                       (2, 5)   
3                                       (2, 5)   
4                                       (2, 3)   

   Section 5 of 6 [I am not interested in abstract ideas.]  \
ID                                                           
0                                              (5, 3)        
1                                              (5, 3)        
2                                              (5, 5)        
3                                              (5, 3)        
4                                              (5, 3)        

   Section 5 of 6 [I am not interested in other people's problems.]  \
ID                                                                    
0                                              (2, 3)                 
1                                              (2, 5)                 
2                                              (2, 5)                 
3                                              (2, 3)                 
4                                              (2, 3)                 

   Section 5 of 6 [I am not really interested in others.]  \
ID                                                          
0                                              (2, 5)       
1                                              (2, 5)       
2                                              (2, 5)       
3                                              (2, 3)       
4                                              (2, 3)       

   Section 5 of 6 [I am quick to understand things.]  \
ID                                                     
0                                             (5, 5)   
1                                             (5, 5)   
2                                             (5, 5)   
3                                             (5, 3)   
4                                             (5, 5)   

   Section 5 of 6 [I am quiet around strangers.]  \
ID                                                 
0                                         (1, 3)   
1                                         (1, 3)   
2                                         (1, 1)   
3                                         (1, 3)   
4                                         (1, 1)   

                    ...                    \
ID                  ...                     
0                   ...                     
1                   ...      

In [16]:
per_scores2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1555 entries, 0 to 1554
Data columns (total 50 columns):
Section 5 of 6 [I am always prepared.]                                      1555 non-null object
Section 5 of 6 [I am easily disturbed.]                                     1555 non-null object
Section 5 of 6 [I am exacting (demanding) in my work.]                      1555 non-null object
Section 5 of 6 [I am full of ideas.]                                        1555 non-null object
Section 5 of 6 [I am interested in people.]                                 1555 non-null object
Section 5 of 6 [I am not interested in abstract ideas.]                     1555 non-null object
Section 5 of 6 [I am not interested in other people's problems.]            1555 non-null object
Section 5 of 6 [I am not really interested in others.]                      1555 non-null object
Section 5 of 6 [I am quick to understand things.]                           1555 non-null object
Section 5 of 6 [I am

* Above we used the .info() method to get information about our dataframe. we can see that our dataframe contains objects.

* Below we check for the type in row 3 of the per_scores1 dataframe.

In [17]:
print(per_scores2.iloc[3].dtype)

object


* The eval() method parses the expression passed to this method and runs python expression (code) within the program. We use the eval() method on the dataframe to convert the values to tuples.

* Below we define the function Total_for_Subscale to let us calculate the subscale sum.

In [18]:
def total_for_subscale(per_scores_per_person, subscale):
    subscale_sum = sum([eval(per_scores_per_person[i])[1] for i in range(50) if eval(per_scores_per_person[i])[0] == subscale])
    return(subscale_sum)

* We shall now apply this function to get the subscale scores.

In [19]:
per_scores2['Extraversion'] = per_scores2.apply(lambda row: total_for_subscale(row, 1), axis=1)

In [20]:
per_scores2['Agreeableness'] = per_scores2.apply(lambda row: total_for_subscale(row, 2), axis=1)

In [21]:
per_scores2['Conscientiousness'] = per_scores2.apply(lambda row: total_for_subscale(row, 3), axis=1)

In [22]:
per_scores2['Neuroticism'] = per_scores2.apply(lambda row: total_for_subscale(row, 4), axis=1)

In [23]:
per_scores2['Intellect'] = per_scores2.apply(lambda row: total_for_subscale(row, 5), axis=1)

* we use .head() to check the new columns in the per_scores2 dataframe.

In [24]:
per_scores2.head(5)

Section 5 of 6 [I am always prepared.]  \
ID                                          
0                                  (3, 5)   
1                                  (3, 5)   
2                                  (3, 5)   
3                                  (3, 5)   
4                                  (3, 3)   

   Section 5 of 6 [I am easily disturbed.]  \
ID                                           
0                                   (4, 5)   
1                                   (4, 5)   
2                                   (4, 3)   
3                                   (4, 5)   
4                                   (4, 5)   

   Section 5 of 6 [I am exacting (demanding) in my work.]  \
ID                                                          
0                                              (3, 5)       
1                                              (3, 5)       
2                                              (3, 3)       
3                                              (3, 3)       
4                                              (3, 3)       

   Section 5 of 6 [I am full of ideas.]  \
ID                                        
0                                (5, 5)   
1                                (5, 5)   
2                                (5, 5)   
3                                (5, 5)   
4                                (5, 3)   

   Section 5 of 6 [I am interested in people.]  \
ID                                               
0                                       (2, 3)   
1                                       (2, 5)   
2                                       (2, 5)   
3                                       (2, 5)   
4                                       (2, 3)   

   Section 5 of 6 [I am not interested in abstract ideas.]  \
ID                                                           
0                                              (5, 3)        
1                                              (5, 3)        
2                                              (5, 5)        
3                                              (5, 3)        
4                                              (5, 3)        

   Section 5 of 6 [I am not interested in other people's problems.]  \
ID                                                                    
0                                              (2, 3)                 
1                                              (2, 5)                 
2                                              (2, 5)                 
3                                              (2, 3)                 
4                                              (2, 3)                 

   Section 5 of 6 [I am not really interested in others.]  \
ID                                                          
0                                              (2, 5)       
1                                              (2, 5)       
2                                              (2, 5)       
3                                              (2, 3)       
4                                              (2, 3)       

   Section 5 of 6 [I am quick to understand things.]  \
ID                                                     
0                                             (5, 5)   
1                                             (5, 5)   
2                                             (5, 5)   
3                                             (5, 3)   
4                                             (5, 5)   

   Section 5 of 6 [I am quiet around strangers.]    ...     \
ID                                                  ...      
0                                         (1, 3)    ...      
1                                         (1, 3)    ...      
2                                         (1, 1)    ...      
3                                         (1, 3)    ...      
4                                         (1, 1)    ...      

   Section 5 of 6 [I sympathize with others' feelings.]  \
ID                                 

## 3. Create a function/functions that takes in subscale total, and for each subscale, creates a new column called "{subscale_name}interpretation" (e.g., conscientiousnessinterpretation) and applies the labels "low", "medium", or "high" according to a person's score on that subscale. 

* Create a new dataframe from the per_scores2 dataframe containing just the Extraversion, Agreeableness, Conscientiousness, Neuroticism, and Intellect columns and name it subscale_scores.

In [25]:
subscale_scores = per_scores2[["Extraversion", "Agreeableness", "Conscientiousness", "Neuroticism", "Intellect"]]

* We take a look at the new dataframe.

In [26]:
subscale_scores.head()

Extraversion  Agreeableness  Conscientiousness  Neuroticism  Intellect
ID                                                                        
0             30             40                 48           36         42
1             42             46                 46           40         42
2             28             40                 40           38         42
3             30             38                 38           40         38
4             28             34                 46           38         36

In [27]:
subscale_scores.describe()

Extraversion  Agreeableness  Conscientiousness  Neuroticism  \
count   1555.000000    1555.000000        1555.000000  1555.000000   
mean      30.852733      42.180064          42.149196    36.118328   
std        8.076902       5.346042           5.924742     7.025337   
min       10.000000      22.000000          16.000000    12.000000   
25%       26.000000      38.000000          38.000000    32.000000   
50%       30.000000      42.000000          44.000000    38.000000   
75%       36.000000      46.000000          46.000000    42.000000   
max       50.000000      50.000000          50.000000    50.000000   

         Intellect  
count  1555.000000  
mean     40.086174  
std       5.469963  
min      20.000000  
25%      36.000000  
50%      40.000000  
75%      44.000000  
max      50.000000

In [28]:
subscale_scores.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1555 entries, 0 to 1554
Data columns (total 5 columns):
Extraversion         1555 non-null int64
Agreeableness        1555 non-null int64
Conscientiousness    1555 non-null int64
Neuroticism          1555 non-null int64
Intellect            1555 non-null int64
dtypes: int64(5)
memory usage: 72.9 KB


In [29]:
import statistics

* Below we use .loc to select the entire dataframe and .mean() and .std() methods to calculate the mean and standard deviation of the dataframe columns.

In [30]:
subscale_scores.loc[ : , : ].mean()

Extraversion         30.852733
Agreeableness        42.180064
Conscientiousness    42.149196
Neuroticism          36.118328
Intellect            40.086174
dtype: float64

In [31]:
subscale_scores.loc[ : , : ].std()

Extraversion         8.076902
Agreeableness        5.346042
Conscientiousness    5.924742
Neuroticism          7.025337
Intellect            5.469963
dtype: float64

Below we define a Score_interpretation function to help us calculate the interpreted scores.

In [32]:
def generate_score_interpretation(score):
# categories = ['low', 'medium', 'high']
# defining function to set the categories
    for columns in subscale_scores.columns:
        # we use the mean and std calculations from above to get the High and Low variables.
        high = subscale_scores.loc[:, columns].mean() + ((subscale_scores.loc[:, columns].std()) / 2)
        low = subscale_scores.loc[:, columns].mean() - ((subscale_scores.loc[:, columns].std()) / 2)
        
        if score < low: 
            return 'low' 

        elif score >= low and score < high: 
            return 'medium'
        
        else :
            return 'high'

* Now we apply the scores_interpretation function to subscale scores and create new columns 

In [33]:
subscale_scores[['Extraversion_Interpretation','Agreeableness_Interpretation','Conscientiousness_Interpretation', 'Neuroticism_Interpretation','Openness_Interpretation']] = subscale_scores.applymap(generate_score_interpretation)
subscale_scores.head()

Extraversion  Agreeableness  Conscientiousness  Neuroticism  Intellect  \
ID                                                                           
0             30             40                 48           36         42   
1             42             46                 46           40         42   
2             28             40                 40           38         42   
3             30             38                 38           40         38   
4             28             34                 46           38         36   

   Extraversion_Interpretation Agreeableness_Interpretation  \
ID                                                            
0                       medium                         high   
1                         high                         high   
2                       medium                         high   
3                       medium                         high   
4                       medium                       medium   

   Conscientiousness_Interpretation Neuroticism_Interpretation  \
ID                                                               
0                              high                       high   
1                              high                       high   
2                              high                       high   
3                              high                       high   
4                              high                       high   

   Openness_Interpretation  
ID                          
0                     high  
1                     high  
2                     high  
3                     high  
4                     high

### Read in the data in departments.csv and merge this data frame with the personality score data frame.

* Use pandas to read in the departments.csv file and call this file departments.

In [34]:
departments = pd.read_csv("departments.csv", delimiter = ";")

* Again we use the .head() method to see what our dataframe looks like.

In [35]:
departments.head()

ID Department  Unnamed: 2  Unnamed: 3  Unnamed: 4  Unnamed: 5  Unnamed: 6  \
0   0       Data         NaN         NaN         NaN         NaN         NaN   
1   1       Data         NaN         NaN         NaN         NaN         NaN   
2   2       Data         NaN         NaN         NaN         NaN         NaN   
3   3       Data         NaN         NaN         NaN         NaN         NaN   
4   4       Data         NaN         NaN         NaN         NaN         NaN   

   Unnamed: 7  Unnamed: 8  Unnamed: 9       ...        Unnamed: 11  \
0         NaN         NaN         NaN       ...                NaN   
1         NaN         NaN         NaN       ...                NaN   
2         NaN         NaN         NaN       ...                NaN   
3         NaN         NaN         NaN       ...                NaN   
4         NaN         NaN         NaN       ...                NaN   

   Unnamed: 12  Unnamed: 13  Unnamed: 14  Unnamed: 15  Unnamed: 16  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   

   Unnamed: 17  Unnamed: 18  Unnamed: 19  IPIP_HIGH_RISK  
0          NaN          NaN          NaN             NaN  
1          NaN          NaN          NaN             NaN  
2          NaN          NaN          NaN             NaN  
3          NaN          NaN          NaN             NaN  
4          NaN          NaN          NaN             NaN  

[5 rows x 21 columns]

* We use the .info() method to see what is our info made of.

In [36]:
departments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1555 entries, 0 to 1554
Data columns (total 21 columns):
ID                1555 non-null int64
Department        1555 non-null object
Unnamed: 2        0 non-null float64
Unnamed: 3        0 non-null float64
Unnamed: 4        0 non-null float64
Unnamed: 5        0 non-null float64
Unnamed: 6        0 non-null float64
Unnamed: 7        0 non-null float64
Unnamed: 8        0 non-null float64
Unnamed: 9        0 non-null float64
Unnamed: 10       0 non-null float64
Unnamed: 11       0 non-null float64
Unnamed: 12       0 non-null float64
Unnamed: 13       0 non-null float64
Unnamed: 14       0 non-null float64
Unnamed: 15       0 non-null float64
Unnamed: 16       0 non-null float64
Unnamed: 17       0 non-null float64
Unnamed: 18       0 non-null float64
Unnamed: 19       0 non-null float64
IPIP_HIGH_RISK    0 non-null float64
dtypes: float64(19), int64(1), object(1)
memory usage: 255.2+ KB


* Create a new dataframe and call it depts. This dataframe will only contain the non-null value containing columns as shown by the above output.

In [37]:
depts = departments[["ID", "Department"]]

In [38]:
depts.head()

ID Department
0   0       Data
1   1       Data
2   2       Data
3   3       Data
4   4       Data

* Now we will merge the depts df and the subscale df, thus creating a merged_df from the two. We shallalso set the ID column from the depts df as the index for the new dataframe.

In [39]:
merged_df = pd.concat([depts, subscale_scores], 1)

In [40]:
merged_df.set_index('ID')
merged_df.head()

ID Department  Extraversion  Agreeableness  Conscientiousness  Neuroticism  \
0   0       Data            30             40                 48           36   
1   1       Data            42             46                 46           40   
2   2       Data            28             40                 40           38   
3   3       Data            30             38                 38           40   
4   4       Data            28             34                 46           38   

   Intellect Extraversion_Interpretation Agreeableness_Interpretation  \
0         42                      medium                         high   
1         42                        high                         high   
2         42                      medium                         high   
3         38                      medium                         high   
4         36                      medium                       medium   

  Conscientiousness_Interpretation Neuroticism_Interpretation  \
0                             high                       high   
1                             high                       high   
2                             high                       high   
3                             high                       high   
4                             high                       high   

  Openness_Interpretation  
0                    high  
1                    high  
2                    high  
3                    high  
4                    high

### Assert that the newly created merged data frame has the same amount of rows as the department data frame, and the expected number of columns. 

* Below we use print(_____.shape) on the departments and the merged_df dataframes to check the number of rows and columns they each have.

In [41]:
print(departments.shape)
print(merged_df.shape)

(1555, 21)
(1555, 12)
